server/src/main/java/org/candlepin/model/HostCache.java

```diff
@@ -0,0 +1,26 @@
+/**
+ * Copyright (c) 2009 - 2018 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.model;
+
+import org.candlepin.guice.CandlepinRequestScoped;
+
+import java.util.HashMap;
+
+/**
+ * Cache for host consumers (see https://github.com/candlepin/candlepin/pull/1334 for original rationale)
+ */
+@CandlepinRequestScoped
```

Comment:
What happens when this cache is used from outside a request scope (i.e. from within the jobs)? Do we get an exception for being out of scope, or do we get a new instance on every call? We could probably work around this by keeping it session scoped rather than request scope in either case.

While I'm thinking on it, this is also an opportunity to make a proper type-safe minicache object that acts like a cache instead of a store. That way we can reduce our memory load in the off chance we flood this cache with hosts in a memory-heavy operation. If not this PR, I feel we should at least throw a task on the backlog to do so.

Reply:
> What happens when this cache is used from outside a request scope (i.e. from within the jobs)?

Javadoc for CandlepinRequestScope says "works even in our Quartz jobs".

> While I'm thinking on it, this is also an opportunity to make a proper type-safe minicache object that acts like a cache instead of a store. That way we can reduce our memory load in the off chance we flood this cache with hosts in a memory-heavy operation. If not this PR, I feel we should at least throw a task on the backlog to do so.

We can certainly look at using the caching mechanisms we introduced in #1336 , but I'd be against building anything too complicated. Created a card to discuss further.